### ISSUE TITLE
Rule proposal: no-array-delete

### ISSUE THREAD

**Issue by davidmaxwaterman** [typescript-eslint/typescript-eslint#4432]
### Description

Deleting an array element can lead to unexpected (for some) behaviour since it leaves 'length' unchanged. Some might say that using delete on array elements should best be avoided and there are alternatives - splice() seems to be preferred.

Having a rule to prevent delete being used on arrays would be beneficial.

### Fail

```ts
> var myArray = [ 1, 2, 3 ];
undefined
> delete myArray[1];
true
> myArray
(3) [1, empty, 3]
> myArray[1]
undefined
> myArray.length
3
```

### Pass

```ts
> var myArray = [ 1, 2, 3 ];
(3) [1, 2, 3]
> myArray.splice(1,1)
[2]
> myArray
(2) [1, 3]
```

**Comment by omril1** [comment]
Also leaving holes in the array makes JS engine remove some optimizations from it.

**Comment by RebeccaStevens** [comment]
This might be overkill for what you are look for but [functional/immutable-data](https://github.com/jonaskello/eslint-plugin-functional/blob/master/docs/rules/immutable-data.md) will prevent this.

**Comment by davidmaxwaterman** [comment]
I've not tried it, but the description of  functional/immutable-data does not seem to include delete for arrays, only for objects (where there isn't an issue).

**Comment by davidmaxwaterman** [comment]
Having said that, and now I'm back at a computer, I can see it does actually flag it as an error:

  3:1  error  Modifying an existing object/array is not allowed  functional/immutable-data

I created a PR to add it to the docs:

https://github.com/jonaskello/eslint-plugin-functional/pull/319

**Comment by davidmaxwaterman** [comment]
@RebeccaStevens Do you know if there's a way to use functional/immutable-data to prevent just delete on arrays? ...or is it all or nothing?

**Comment by RebeccaStevens** [comment]
Currently it is all or nothing. Options can be added to extend the functionality.

**Comment by NanderTGA** [comment]
+1 One of my own projects and a similar one encountered issues because of deleting values from an array like this.
Great to see that it can be prevented in the future.
